File: lib/src/broadcast_center.dart
import 'dart:ffi' as ffi;
import 'package:ffi/ffi.dart';

import 'package:nativeapi/src/broadcast_receiver.dart';
import 'package:nativeapi/src/ffi/bindings_generated.dart';
import 'ffi/bindings.dart';

class BroadcastCenter {
  BroadcastCenter._() {
    _broadcastReceivedCallbackCallable =
        ffi.NativeCallable<BroadcastReceivedCallbackFunction>.listener(
            _onBroadcastReceived);
    _bindings.broadcast_center_on_broadcast_received(
        _broadcastReceivedCallbackCallable.nativeFunction);
  }

  static final BroadcastCenter instance = BroadcastCenter._();

  NativeApiBindings get _bindings => nativeApiBindings;

  late final ffi.NativeCallable<BroadcastReceivedCallbackFunction>
      _broadcastReceivedCallbackCallable;

  final Map<String, List<BroadcastReceiver>> _receivers = {};

  void _onBroadcastReceived(
    ffi.Pointer<ffi.Char> topic,
    ffi.Pointer<ffi.Char> message,
  ) {
    String topicStr;
    String messageStr;

    try {
      topicStr = topic.cast<Utf8>().toDartString();
    } catch (e) {
      print('Error converting topic to string: $e');
      topicStr = 'invalid_topic';
    }

    try {
      messageStr = message.cast<Utf8>().toDartString();
    } catch (e) {
      print('Error converting message to string: $e');
      messageStr = 'invalid_message';
    }

    print('onBroadcastReceived: $topicStr, $messageStr');
    notifyReceivers(
      topicStr,
      (l) => l.onBroadcastReceived(
        topicStr,
        messageStr,
      ),
    );
  }

  /// Send a broadcast message to all receivers of a given topic
  void sendBroadcast(String topic, String message) {
    _bindings.broadcast_center_send_broadcast(
      topic.toNativeUtf8().cast<ffi.Char>(),
      message.toNativeUtf8().cast<ffi.Char>(),
    );
  }

  /// Register a receiver for a given topic
  void registerReceiver(String topic, BroadcastReceiver listener) {
    if (_receivers[topic] == null) {
      _receivers[topic] = [];
    }
    _receivers[topic]?.add(listener);
    _bindings.broadcast_center_register_receiver(
      topic.toNativeUtf8().cast<ffi.Char>(),
    );
  }

  /// Unregister a receiver for a given topic
  void unregisterReceiver(String topic, BroadcastReceiver listener) {
    _receivers[topic]?.remove(listener);
    _bindings.broadcast_center_unregister_receiver(
      topic.toNativeUtf8().cast<ffi.Char>(),
    );
  }

  /// Notify all receivers of a given topic and message
  ///
  /// The [topic] is the topic of the broadcast.
  /// The [callback] is called for each receiver.
  void notifyReceivers(String topic, Function(BroadcastReceiver) callback) {
    print('notifyReceivers: $topic');

    final firstTopic = _receivers.keys.first.toString();

    for (var listener in _receivers[firstTopic] ?? []) {
      print('notifyReceivers: $topic, $listener');
      callback(listener);
    }
  }
}

File: src/broadcast_center.cpp
#include <string.h>
#include <iostream>

#include "broadcast_center.h"
#include "libnativeapi/include/nativeapi.h"

using namespace nativeapi;

static BroadcastCenter g_broadcast_center = BroadcastCenter();
static std::map<std::string, std::unique_ptr<BroadcastEventHandler>>
    g_broadcast_event_handlers;
static BroadcastReceivedCallback g_broadcast_received_callback = nullptr;

FFI_PLUGIN_EXPORT
void broadcast_center_send_broadcast(const char* topic, const char* message) {
  g_broadcast_center.SendBroadcast(topic, message);
}

FFI_PLUGIN_EXPORT
void broadcast_center_register_receiver(const char* topic) {
  std::unique_ptr<BroadcastEventHandler> broadcast_event_handler =
      std::make_unique<BroadcastEventHandler>(
          [](const std::string& topic, const std::string& message) {
            if (g_broadcast_received_callback) {
              char* topic_cstr = strdup(topic.c_str());
              char* message_cstr = strdup(message.c_str());
              g_broadcast_received_callback(topic_cstr, message_cstr);
            }
          });
  g_broadcast_event_handlers[topic] = std::move(broadcast_event_handler);
  g_broadcast_center.RegisterReceiver(topic,
                                      g_broadcast_event_handlers[topic].get());
}

FFI_PLUGIN_EXPORT
void broadcast_center_unregister_receiver(const char* topic) {
  g_broadcast_center.UnregisterReceiver(
      topic, g_broadcast_event_handlers[topic].get());
  g_broadcast_event_handlers.erase(topic);
}

FFI_PLUGIN_EXPORT
void broadcast_center_on_broadcast_received(
    BroadcastReceivedCallback callback) {
  g_broadcast_received_callback = callback;
}
